tlvc/lib/heap.c

```diff
@@ -150,8 +155,34 @@ void heap_free(void *data)
 	//     chance on next allocation of same size
 
 	list_s *i = &_heap_free;
+	list_s *n;
 	debug_heap("free 0x%x/%u; ", h, h->size);
 
+#ifdef HEAPFREE_VALIDATE_1
+	/* Sanity check */
+	heap_s *this;
+	n = _heap_all.prev;
+	while (n != &_heap_all) {
+		this = structof(n, heap_s, all);
+		//printk("this+1 %x data %x tag %x\n", this+1, data, this->tag);
+		if ((this->tag & HEAP_TAG_USED) && data == (this+1))
+			break;
+		n = n->prev;	/* going backwards is slightly faster */
+	}
+	if (n == &_heap_all) {
+		printk("WARNING: bogus heap_free");
+		return;
+	}
+#endif
+
+#ifdef  HEAPFREE_VALIDATE_2
+	//printk("Heap canary %x\n", *((char *)data - 1)&0xff);
+	if ((*((char *)data - 1)&0xff) != HEAP_CANARY) {
```

Comment:
Can't you just use `if (h->canary != HEAP_CANARY)` here rather than using complicated pointer arithmetic that also assumes canary is the last struct member?

Reply:
Not my day! Thanks.